single_kernel_mongo/events/database.py

```diff
@@ -0,0 +1,138 @@
+    """Manager for handling database events."""
+
+    def __init__(self, dependent: OperatorProtocol, relation_name: RelationNames):
+        super().__init__(parent=dependent, key=relation_name.value)
+        self.dependent = dependent
+        self.manager = dependent.mongo_manager
+        self.charm = dependent.charm
+        self.relation_name = relation_name
+        self.database_provides = DatabaseProvides(self.charm, relation_name=self.relation_name)
+
+        self.framework.observe(
+            self.charm.on[relation_name].relation_departed,
+            self.dependent.check_relation_broken_or_scale_down,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_broken, self._on_relation_event
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_changed, self._on_relation_event
+        )
+        self.framework.observe(
+            self.database_provides.on.database_requested, self._on_relation_event
+        )
+
+    def _on_relation_event(self, event: RelationEvent):
+        """Handle relation joined events.
+
+        When relations join, change, or depart, the :class:`MongoDBClientRelation`
+        creates or drops MongoDB users and sets credentials into relation
+        data. As a result, related charm gets credentials for accessing the
+        MongoDB database.
+        """
+        relation_departing = False
+        relation_changed = False
+        # TODO : Handle the Mongos VM case
+        # (https://github.com/canonical/mongos-operator/blob/6/edge/lib/charms/mongos/v0/mongos_client_interface.py)
+        # for interface unification.
+        try:
+            if not self.pass_hook_checks(event):
+                logger.info(f"Skipping {type(event)}: Hook checks did not pass")
+                return
+        except UpgradeInProgressError:
+            logger.warning(
+                "Adding relations is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+            logger.info(f"Skipping {type(event)}: Hook checks did not pass")
+            event.defer()
+            return
+
+        if isinstance(event, RelationBrokenEvent):
+            relation_departing = True
+            # TODO: Checks
+            if not self.dependent.state.has_departed_run(event.relation.id):
+                defer_event_with_info_log(
+                    logger,
+                    event,
+                    "relation broken",
+                    "must wait for relation departed hook to decide if relation should be removed.",
+                )
+                return
+            if self.dependent.state.is_scaling_down(event.relation.id):
+                logger.info(
+                    "Relation broken event due to scale down, do not proceed to remove users."
+                )
+                return
+            logger.info("Relation broken event due to relation removal, proceed to remove user.")
```

Comment:
Ah - maybe we can have a quick sync on this in our next 1:1. I am a little lost 😅 

The question is more that every relation should check if the relation is broken or scaling down, so it might make sense to have the functionality added everywhere (this doesn't have to be done in this PR but we shouldn't forget about it)

Reply:
There are two things:
 * Not all relations handle the relation departed event the same way currently and we need to ensure that it's safe to unify it (and maybe it is, but we need to be sure)
 *  We would need a small rewrite so that this new function raises a `DeferrableError` in the branch `has_departed_run` so that the final handler can raise

Reply:
It looks like this was moved in one of your subsequent PRs and is now raising `DeferrableError `?

Reply:
It is very plausible that I did that and did not remember :grin: